### PR TITLE
[SPARK-23519][SQL] create view should work from query with duplicate output columns

### PR DESCRIPTION
**Pull request by ghost**
**What changes were proposed in this pull request?**

Moving the call for checkColumnNameDuplication out of generateViewProperties. This way we can choose ifcheckColumnNameDuplication will be performed on analyzed or aliased plan without having to pass an additional argument(aliasedPlan) to generateViewProperties.

Before the pr column name duplication was performed on the query output of below sql(c1, c1) and the pr makes it perform check on the user provided schema of view definition(c1, c2)

**Why are the changes needed?**

Changes are to fix SPARK-23519 bug. Below queries would cause an exception. This pr fixes them and also added a test case.

`CREATE TABLE t23519 AS SELECT 1 AS c1
CREATE VIEW v23519 (c1, c2) AS SELECT c1, c1 FROM t23519`

**Does this PR introduce any user-facing change?**
No

**How was this patch tested?**
new unit test added in SQLViewSuite

### CODE REVIEW

**Comment by viirya** [comment]
I'd suggest to use predefined PR description template. cc @HyukjinKwon

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 1]
```diff
@@ -363,11 +374,6 @@ object ViewHelper {
       analyzedPlan: LogicalPlan): Map[String, String] = {
     val queryOutput = analyzedPlan.schema.fieldNames
 
-    // Generate the query column names, throw an AnalysisException if there exists duplicate column
-    // names.
-    SchemaUtils.checkColumnNameDuplication(
-      queryOutput, "in the view definition", session.sessionState.conf.resolver)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Why not make change here? Let checkColumnNameDuplication check the output of aliasedPlan?

**Review by viirya — COMMENTED** [review]
Does alter view as command also have this issue?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 2]
```diff
@@ -710,4 +710,10 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("SPARK-23519 view should be created even when query output contains duplicate col name") {
+    sql("CREATE TABLE t23519 AS SELECT 1 AS c1")
```
[the remark this answers is not in the document]
Comment (hem1891) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: My bad. Corrected ,Thanks

**Comment by hem1891** [comment]
> have

ALTER VIEW AS gets its schema from the new SELECT query provided. So it is not effected

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
ok to test

**Comment by viirya** [comment]
> > have
> 
> ALTER VIEW AS gets its schema from the new SELECT query provided. So it is not effected

But `sql("ALTER VIEW v23519 AS SELECT c1, c1 FROM t23519")` also fails with an exception?

```scala
org.apache.spark.sql.AnalysisException: Found duplicate column(s) in the view definition: `c1`;
[info]   at org.apache.spark.sql.util.SchemaUtils$.checkColumnNameDuplication(SchemaUtils.scala:90)                                                   
[info]   at org.apache.spark.sql.util.SchemaUtils$.checkColumnNameDuplication(SchemaUtils.scala:70)                                                   
[info]   at org.apache.spark.sql.execution.command.ViewHelper$.generateViewProperties(views.scala:369)
[info]   at org.apache.spark.sql.execution.command.AlterViewAsCommand.alterPermanentView(views.scala:301)
```

**Comment by maropu** [comment]
oh, I see...

**Comment by hemanthmeka** [comment]
> > ALTER VIEW AS gets its schema from the new SELECT query provided. So it is not effected
> 
> But `sql("ALTER VIEW v23519 AS SELECT c1, c1 FROM t23519")` also fails with an exception?
> 
> ```scala
> org.apache.spark.sql.AnalysisException: Found duplicate column(s) in the view definition: `c1`;
> [info]   at org.apache.spark.sql.util.SchemaUtils$.checkColumnNameDuplication(SchemaUtils.scala:90)                                                   
> [info]   at org.apache.spark.sql.util.SchemaUtils$.checkColumnNameDuplication(SchemaUtils.scala:70)                                                   
> [info]   at org.apache.spark.sql.execution.command.ViewHelper$.generateViewProperties(views.scala:369)
> [info]   at org.apache.spark.sql.execution.command.AlterViewAsCommand.alterPermanentView(views.scala:301)
> ```

From the comments in CreateViewCommand.run('**Nothing we need to retain from the old view, so ...**') i get the impression that alter always creates a new definition with schema from query, then c1, c1 should throw analysis exception. Which should be the right behaviour? Any reference?

**Comment by viirya** [comment]
hmm I am not sure if ALTER VIEW AS should take the schema of the new query, or it should keep original schema, and just replace old query?

That's right, you described is what ALTER VIEW AS works now, as the same as CREATE OR REPLACE VIEW: the schema comes from new query and replaces old one. Hive's document https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-AlterViewAsSelect, doesn't talk the details, just says it works as CREATE OR REPLACE VIEW. But I can't find Hive doc of CREATE OR REPLACE VIEW. I assume it works like Spark SQL.

Postgresql has CREATE OR REPLACE VIEW, but interestingly, the new query in CREATE OR REPLACE VIEW must generate same column names in the same order and data types. No ALTER VIEW AS in Postgresql.

I thought ALTER VIEW AS only replaces old query, and keeps original schema. But seems it isn't how ALTER VIEW AS works now.

**Comment by viirya** [comment]
Btw, in the original JIRA comment, someone claims that Hive supports duplicate column names. I want to confirm but get:

```
hive> create view test_view (c1, c2) as select c1, c1 from test;
FAILED: SemanticException [Error 10036]: Duplicate column name: c1
```

I test with Hive 2.1. Does only newer Hive support the feature?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 3]
```diff
@@ -360,13 +362,15 @@ object ViewHelper {
...
       session: SparkSession,
-      analyzedPlan: LogicalPlan): Map[String, String] = {
+      analyzedPlan: LogicalPlan,
+      fieldNames: Array[String]): Map[String, String] = {
+    // for createViewCommand queryOutput may be different from fieldNames
     val queryOutput = analyzedPlan.schema.fieldNames
 
     // Generate the query column names, throw an AnalysisException if there exists duplicate column
     // names.
     SchemaUtils.checkColumnNameDuplication(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: we can make the diff much smaller by changing this line to
```
val colNamesToCheck = if (userSpecifiedColumns.nonEmpty) userSpecifiedColumns.map(_._1) else queryOutput
SchemaUtils.checkColumnNameDuplication(colNamesToCheck, ...)
```

Comment (hem1891) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: > we can make the diff much smaller by changing this line to
> 
> ```
> val colNamesToCheck = if (userSpecifiedColumns.nonEmpty) userSpecifiedColumns.map(_._1) else queryOutput
> SchemaUtils.checkColumnNameDuplication(colNamesToCheck, ...)
> ```
userSpecifiedColumns is not accessible in SchemaUtils. I guess we will have to pass them as arguments

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: We only need to access `userSpecifiedColumns` before calling `SchemaUtils`, don't we?

Comment (hem1891) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: My bad, I meant `userSpecifiedColumns` is not available in `ViewHelper` not `SchemaUtils`

- `CreateViewCommand.prepareTable()` and `AlterViewAsCommand.alterPermanentView()` both call  `ViewHelper.generateViewProperties()`
- And `ViewHelper.generateViewProperties()` is where we call `SchemaUtils.checkColumnNameDuplication()`. The suggestion you provided requires `userSpecifiedColumns` to be accessible in `ViewHelper.generateViewProperties()`
- So i added an argument to `ViewHelper.generateViewProperties()` allowing `CreateViewCommand` to send aliased schema and `AlterViewAsCommand` to send analyzed schema

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by gatorsmile** [comment]
@hem1891 Could you open a PR against 2.4? We can backport it.

**Comment by hemanthmeka** [comment]
> @hem1891 Could you open a PR against 2.4? We can backport it.

sure

**Comment by hem1891** [comment]
> @hem1891 Could you open a PR against 2.4? We can backport it.

PR raised. [25733](https://github.com/apache/spark/pull/25733)

**Pull request closed without merging**